pep-0626.rst

```diff
@@ -182,6 +182,57 @@ Access to the ``f_lineno`` attribute of frame objects through C API functions is
 
 Accessing ``f_lineno`` directly through the underlying data structure is forbidden.
 
+Out of process debuggers and profilers
+''''''''''''''''''''''''''''''''''''''
+
+Out of process tools, such as py-spy [1]_, cannot use the C-API, and must parse the line number table themselves.
+Although the line number table format may change without warning,
+it will not change during a release unless absolutely necessary for a bug fix.
+
+To reduce the work required to implement these tools, the following C struct and utility functions are provided.
+Note that these functions are not part of the C-API, so will be need to be linked into any code that needs to use them.
+
+::
+
+    typedef struct addressrange {
+        int ar_start;
+        int ar_end;
+        int ar_line;
+        int opaque1;
+        void *opaque2;
```

Comment:
This only works if `PyLineTable_InitAddressRange` and `PyLineTable_InitAddressRange` does not itself access the `opaque2` as this pointer will be in the remote process, not in the profiler one.

Reply:
Now you're just making stuff up, to object to.
Why would you assume that a pointer would point to invalid memory?

Reply:
Unless, of course, you pass an invalid pointer to `PyLineTable_InitAddressRange`. Not much I can do about that.

Reply:
I am not objecting, I do like the proposal, I am just trying to make sure this mechanism will work as intended.

> Now you're just making stuff up, to object to.
> Why would you assume that a pointer would point to invalid memory?

If you copy the structure from process A to process B, then the opaque pointer in the structure will contain a memory address that refers to process A. Trying to dereference that pointer in process B will segfault because that address is invalid in process B. In order to make it work you need to copy the structure pointed by that pointer from the memory of process A. 

This means, when executing the function in process B you would need to substitute `opaque_pointer->some_attr` with `copy_memory(opaque_ptr, num_bytes, &dest); dest->some_attr`.

Reply:
To make this more clear, let's say you have process A and process B. The structure in process A (the python process) looks like this (not the ex:

{ 1, 2, 3, 0xa12345678 }

The memory address that corresponds to the opaque pointer, 0xa12345678 is valid and works fine in process A because that is where is intended to be used. Dereferencing this pointer using `opaque->attr` will work.

Now you copy this struct to process B. In process B the memory address 0xa12345678 is invalid because it points to something in process A. Trying to treat this as a pointer and doing `opaque->attr` will segfault. What process B needs to do is copy from process A whatever is pointed by the address 0xa12345678 and then accessing the field `attr` from the copied memory.

With this in mind, what I am trying to explain of that if these functions that we are proposing cannot dereference pointers if we intend people to use them 'as it is' in their profilers/debuggers.

Reply:
All `PyCodeAddressRange` structs would be in the debugger.
Why would it be in the Python process? That makes no sense.

Reply:
> All `PyCodeAddressRange` structs would be in the debugger.
> Why would it be in the Python process? That makes no sense.

Because that's where the debugger copied them from, no?

Reply:
When the debugger copies the structs from the Python process, they live in the debugger, yes, but the values of the pointers that were copied still point to memory on the Python process.

Reply:
The line number table, doesn't contain any pointers, just bytes representing offsets. There are no pointers.
The `PyCodeAddressRange` is just a structure for iterating over the line number table. There is no concept of local or remote here, just plain old C code.

Reply:
> The `PyCodeAddressRange` is just a structure for iterating over the line number table. There is no concept of local or remote here, just plain old C code.

Ohhhhhhh! So you are saying that the `PyLineTable_InitAddressRange` and `PyLineTable_NextAddressRange` won't access the `opaque2` pointer in the `PyCodeAddressRange` struct?

Reply:
No. `PyLineTable_InitAddressRange` initializes it, and `PyLineTable_NextAddressRange` uses. It's marked opaque, so that user code shouldn't modify it, or assume that it's value is meaningful.

I'm genuinely perplexed why of all the many millions of pointers in C code, you think this one will be so problematic?

If you pass invalid values to `PyLineTable_InitAddressRange`, then either it or `PyLineTable_NextAddressRange` will crash or return incorrect results. Pass valid data and it will work (assuming it is bug free).  Just like all C code.

Reply:
> I'm genuinely perplexed why of all the many millions of pointers in C code, you think this one will be so problematic?

I think the confusion here is that I was assuming that the `PyCodeAddressRange` struct was something that the debugger will need to copy from the Python process and therefore the pointers would be pointing to memory in the remote, but (correct me if I am wrong) it seems that this is something that lives entirely on the debugger and is only used for iteration, no?

Reply:
Yes. In the example, `PyCodeAddressRange range` is allocated on the debugger's stack.

Reply:
Fantastic, all is good now! Thanks for the patience, @markshannon :)

Reply:
It's rare that I'm called patient :smile: 
Thanks for the thorough review.